ports/psoc6/machine_uart.c

```diff
@@ -0,0 +1,509 @@
+// mpy includes
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "py/ringbuf.h"
+#include "py/runtime.h"
+#include "shared/runtime/interrupt_char.h"
+#include "shared/runtime/mpirq.h"
+
+// MTB includes
+#include "cybsp.h"
+#include "cyhal.h"
+
+// port specific includes
+#include "modmachine.h"
+#include "machine_pin_phy.h"
+#include "mplogger.h"
+
+#define UART_HWCONTROL_RTS  (1)
+#define UART_HWCONTROL_CTS  (2)
+#define MAX_UART             10 // TODO: Change as per bsp?
+#define DEFAULT_UART_PIN    (MP_ROM_QSTR(MP_QSTR_NC))
+
+// OR-ed IRQ flags which are allowed to be used by the user
+#define MP_UART_ALLOWED_FLAGS (CYHAL_UART_IRQ_TX_EMPTY | CYHAL_UART_IRQ_TX_DONE | CYHAL_UART_IRQ_RX_FULL | CYHAL_UART_IRQ_RX_DONE)
+
+
+#define uart_assert_raise_val(msg, ret)   if (ret != CY_RSLT_SUCCESS) { \
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT(msg), ret); \
+}
+
+typedef struct _machine_uart_obj_t {
+    mp_obj_base_t base;
+    cyhal_uart_t uart_obj;
+    uint32_t init_flag;    // Flag to support reinitialisation of uart parameters
+    uint32_t bits;
+    uint32_t stop;
+    uint32_t baudrate;
+    cyhal_uart_parity_t parity;
+    machine_pin_phy_obj_t *tx;
+    machine_pin_phy_obj_t *rx;
+    machine_pin_phy_obj_t *cts;
+    machine_pin_phy_obj_t *rts;
+    cyhal_uart_cfg_t cfg;
+    uint16_t timeout; // only used by cyhal_uart_getc() ie, readchar()
+    uint8_t flow;
+    uint32_t rxbuf;
+    bool interrupt;
+    uint16_t mp_irq_trigger;            // user IRQ trigger mask
+    uint16_t mp_irq_flags;              // user IRQ active IRQ flags
+    mp_irq_obj_t *mp_irq_obj;           // user IRQ object
+} machine_uart_obj_t;
+
+#define MICROPY_PY_MACHINE_UART_CLASS_CONSTANTS \
+    { MP_ROM_QSTR(MP_QSTR_RTS), MP_ROM_INT(UART_HWCONTROL_RTS) }, \
+    { MP_ROM_QSTR(MP_QSTR_CTS), MP_ROM_INT(UART_HWCONTROL_CTS) }, \
+    { MP_ROM_QSTR(MP_QSTR_TX_EMPTY), MP_ROM_INT(CYHAL_UART_IRQ_TX_EMPTY) }, \
+    { MP_ROM_QSTR(MP_QSTR_TX_DONE), MP_ROM_INT(CYHAL_UART_IRQ_TX_DONE) }, \
+    { MP_ROM_QSTR(MP_QSTR_RX_FULL), MP_ROM_INT(CYHAL_UART_IRQ_RX_FULL) }, \
+    { MP_ROM_QSTR(MP_QSTR_RX_DONE), MP_ROM_INT(CYHAL_UART_IRQ_RX_DONE) }, \
+
+machine_uart_obj_t *uart_obj[MAX_UART] = { NULL };
+
+static const char *_parity_name[] = {"None", "0", "1"};
+
+
+void uart_event_handler(void *handler_arg, cyhal_uart_event_t event) {
```

Comment:
these 2 functions uart_event_handler and uart_irq_config could also be static functions right?